src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,564 @@
+          if (is_correlated) {
+            return LQPVisitation::DoNotVisitInputs;
+          }
+
+          is_correlated |= uses_correlated_parameters(sub_node, is_correlated_parameter);
+          return is_correlated ? LQPVisitation::DoNotVisitInputs : LQPVisitation::VisitInputs;
+        });
+      } else if (sub_expression->type == ExpressionType::CorrelatedParameter) {
+        const auto& parameter_expression = std::static_pointer_cast<CorrelatedParameterExpression>(sub_expression);
+        is_correlated |= is_correlated_parameter(parameter_expression->parameter_id);
+      }
+
+      return is_correlated ? ExpressionVisitation::DoNotVisitArguments : ExpressionVisitation::VisitArguments;
+    });
+
+    if (is_correlated) {
+      break;
+    }
+  }
+
+  return is_correlated;
+}
+
+/**
+ * Searches for usages of correlated parameters that we cannot optimize.
+ *
+ * This includes two things:
+ *   - Usages of correlated parameters outside of predicate nodes (for example joins)
+ *   - Usages of correlated parameters in predicate nodes nested below nodes they cannot be pulled up past.
+ */
+template <class ParameterPredicate>
+bool contains_unoptimizable_correlated_parameter_usages(
+    const std::shared_ptr<AbstractLQPNode>& lqp, ParameterPredicate&& is_correlated_parameter,
+    const std::set<std::shared_ptr<PredicateNode>> optimizable_predicate_nodes) {
+  bool optimizable = true;
+  visit_lqp(lqp, [&](const auto& node) {
+    if (!optimizable) {
+      return LQPVisitation::DoNotVisitInputs;
+    }
+
+    auto correlated = uses_correlated_parameters(node, is_correlated_parameter);
+    if (correlated) {
+      if (node->type != LQPNodeType::Predicate ||
+          optimizable_predicate_nodes.find(std::static_pointer_cast<PredicateNode>(node)) ==
+              optimizable_predicate_nodes.end()) {
+        optimizable = false;
+        return LQPVisitation::DoNotVisitInputs;
+      }
+    }
+
+    return LQPVisitation::VisitInputs;
+  });
+
+  return !optimizable;
+}
+
+/**
+ * Replace aggregate nodes by ones with additional group by expressions added to allow predicates to be pulled up.
+ */
+void replace_aggregate_nodes(PredicatePullUpInfo& pull_up_info) {
+  // Walk the predicates from bottom to top while keeping track of the column expressions they need to be available.
+  // Whenever an aggregate node is encountered, replace it by a new one with the column expressions of all the
+  // predicates below it added as group by expressions.
+
+  // TODO(anybody): This code does quite a bit of unnecessary copying around because the AggregateNode constructor
+  // expects two expression vectors (which it then copies again).
+  std::vector<std::shared_ptr<AbstractExpression>> group_by_expressions;
+  auto aggregate_it = pull_up_info.aggregate_nodes.rbegin();
+  auto aggregate_it_end = pull_up_info.aggregate_nodes.rend();
+  for (auto predicate_idx = pull_up_info.predicates.size(); predicate_idx--;) {
```

Comment:
This... does not look like a nice for loop. Few people expect the middle part to have side effects.

Reply:
just use reverse iterators